### PR TITLE
API server self request metric

### LINKED ISSUE #90908: Add a metric for apiserver self-requests
Currently we don't have an easy way of determining what number of requests are apiserver self-requests. This has historically bitten us, since we used to have a 50(?) QPS limit on apiserver self-requests (which thus has been removed). Even so, it would be a good idea to know this generally, since this may be helpful in pinpointing performance regressions for apiserver.

/sig api-machinery instrumentation

### PR DESCRIPTION
**Pull request by LogicalShark**
/kind feature
/sig api-machinery instrumentation

Adds self-request metric for the API server

Fixes https://github.com/kubernetes/kubernetes/issues/90908

```release-note
Added counter metric "apiserver_request_self" to count API server self-requests with labels for verb, resource, and subresource.
```

### CODE REVIEW

**Comment by LogicalShark** [comment]
@logicalhan

File: staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go [hunk 1]
```diff
@@ -65,3 +53,21 @@ func (a *Authenticator) AuthenticateRequest(req *http.Request) (*authenticator.R
 
 	return resp, ok, err
 }
+
+// ExtractBearerToken gets the bearer token of a request
+func ExtractBearerToken(req *http.Request) (token string, ok bool) {
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go:58]: This looks unit-testable.

File: staging/src/k8s.io/apiserver/pkg/server/servermetrics.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/server/servermetrics.go:22]: We don't actually use prometheus directly. This would have to be registered to our wrapper registry (https://github.com/kubernetes/kubernetes/blob/eda07adf6e3ef8d066ba95a9b1039a13d2e3f8ea/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L266, for instance).

**Review by logicalhan — COMMENTED** [review]
What is the rationale for not using the existing filter/entrypoint? Specifically I'm thinking `k8s.io/apiserver/pkg/endpoints/metrics/metrics.go`.

File: staging/src/k8s.io/apiserver/pkg/server/servermetrics.go (hunk 2 above)
Comment (LogicalShark) [reply staging/src/k8s.io/apiserver/pkg/server/servermetrics.go:22]: Should this documentation be changed? https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/server/servermetrics.go:22]: Yes, it should actually. Would you mind filing an issue so that we don't lose track of it?

Comment (LogicalShark) [reply staging/src/k8s.io/apiserver/pkg/server/servermetrics.go:22]: Done https://github.com/kubernetes/community/issues/5080

File: staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go (hunk 1 above)
Comment (LogicalShark) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go:58]: Added

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics.go [hunk 3]
```diff
@@ -113,3 +133,18 @@ func compressUsername(username string) string {
 		return "other"
 	}
 }
+
+// splitSubresource checks if the given storage path is the path of a subresource and returns
+// the resource and subresource components.
+func splitSubresource(path string) (string, string, error) {
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics.go:139]: this also looks unit-testable

**Review by logicalhan — COMMENTED** [review]
This is looking a lot better! I think we're getting close! Thanks for the patience during the iterations!

File: staging/src/k8s.io/apiserver/go.mod [hunk 4]
```diff
@@ -27,6 +27,7 @@ require (
 	github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822
 	github.com/pkg/errors v0.9.1
 	github.com/pquerna/cachecontrol v0.0.0-20171018203845-0dec1b30a021 // indirect
+	github.com/prometheus/client_golang v1.7.1
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/go.mod:30]: This probably shouldn't be here.

File: cmd/kube-scheduler/app/server.go [hunk 5]
```diff
@@ -220,7 +220,7 @@ func buildHandlerChain(handler http.Handler, authn authenticator.Request, authz
 	failedHandler := genericapifilters.Unauthorized(scheme.Codecs)
 
 	handler = genericapifilters.WithAuthorization(handler, authz, scheme.Codecs)
-	handler = genericapifilters.WithAuthentication(handler, authn, failedHandler, nil)
+	handler = genericapifilters.WithAuthentication(handler, authn, failedHandler, nil, genericapifilters.NoopCheck)
```
Comment (logicalhan) [inline cmd/kube-scheduler/app/server.go:223]: this is inconsistent with `cmd/controller-manager/app/serve.go` which passes nil. 

Since you're handling the NoOp thing in `WithAuthentication`, it probably makes more sense to pass nil here.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication_test.go [hunk 6]
```diff
@@ -91,6 +92,7 @@ func TestAuthenticateRequestWithAud(t *testing.T) {
 					}
 				}),
 				tc.apiAuds,
+				NoopCheck,
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication_test.go:95]: nit: nil for consistency

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication_test.go [hunk 7]
```diff
@@ -128,6 +130,7 @@ func TestAuthenticateRequest(t *testing.T) {
 			t.Errorf("unexpected call to failed")
 		}),
 		nil,
+		NoopCheck,
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication_test.go:133]: nit: nil for consistency

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics.go [hunk 8]
```diff
@@ -90,6 +100,16 @@ func recordAuthMetrics(resp *authenticator.Response, ok bool, err error, apiAudi
 
 	authenticatedAttemptsCounter.WithLabelValues(resultLabel).Inc()
 	authenticationLatency.WithLabelValues(resultLabel).Observe(time.Since(authStart).Seconds())
+
+	if isAPISelfRequest {
+		// When to use canonicalVerb from apiserver/pkg/endpoints/metrics.go?
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics.go:105]: I would use `canonicalVerb` here.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go [hunk 9]
```diff
@@ -146,6 +146,7 @@ func TestMetrics(t *testing.T) {
 					close(done)
 				}),
 				tt.apiAudience,
+				NoopCheck,
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go:149]: nit: nil for consistency

**Review by logicalhan — COMMENTED** [review]
It's almost there! Thanks for iterating over this!

**Comment by logicalhan** [comment]
FYI you are still failing the CLA check. 

```If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address. Check your existing CLA data and verify that your email is set on your git commits.```

**Review by logicalhan — COMMENTED** [review]
ping? it's looking good, just a couple of things plus a squash and it should be ready for review.

**Comment by logicalhan** [comment]
> ```
> Added a metric to count API server self-requests.
> ```

Would you mind adding the literal metric name and the labels here?

Also, can you make this PR no longer WIP/draft state?

File: staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go [hunk 10]
```diff
@@ -65,3 +53,21 @@ func (a *Authenticator) AuthenticateRequest(req *http.Request) (*authenticator.R
 
 	return resp, ok, err
 }
+
+// ExtractBearerToken gets the bearer token of a request, "ok" will be false if the token is missing
+func ExtractBearerToken(req *http.Request) (token string, ok bool) {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go:58]: Can you send this move as a separate PR? It's not good to mix refactoring and functionality change.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 11]
```diff
@@ -27,27 +27,46 @@ import (
...
+func NoopCheck(req *http.Request) bool {
+	return false
+}
+
 // WithAuthentication creates an http handler that tries to authenticate the given request as a user, and then
 // stores any such user found onto the provided context for the request. If authentication fails or returns an error
 // the failed handler is used. On success, "Authorization" header is removed from the request and handler
 // is invoked to serve the request.
-func WithAuthentication(handler http.Handler, auth authenticator.Request, failed http.Handler, apiAuds authenticator.Audiences) http.Handler {
+func WithAuthentication(handler http.Handler, auth authenticator.Request, failed http.Handler, apiAuds authenticator.Audiences, apiSelfRequestCheck func(*http.Request) bool, requestInfoFactory *apirequest.RequestInfoFactory) http.Handler {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:45]: When I said the authentication stuff knows this, I was not expecting you to modify this method, I actually don't recommend this, sorry for not being clear!

Instead, I was expecting a wrapper around the specific authenticator that handles the self requests, that passes through requests and increments the metric if that authenticator returns true.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 12]
```diff
@@ -786,3 +787,12 @@ func AuthorizeClientBearerToken(loopback *restclient.Config, authn *Authenticati
 	tokenAuthorizer := authorizerfactory.NewPrivilegedGroups(user.SystemPrivilegedGroup)
 	authz.Authorizer = authorizerunion.New(tokenAuthorizer, authz.Authorizer)
 }
+
+// IsSelfRequest returns whether req is a self-request on the api server
+func (c *Config) IsSelfRequest(req *http.Request) bool {
+	tok, ok := bearertoken.ExtractBearerToken(req)
+	if ok && (tok == c.LoopbackClientConfig.BearerToken) {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:794]: Yeah, don't do this here. Look for where an authenticator is constructed to match this token instead.
